cmd/backup-manager/app/backup/backup.go

```diff
@@ -0,0 +1,224 @@
+// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package backup
+
+import (
+	"database/sql"
+	"fmt"
+	"io/ioutil"
+	"os/exec"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/golang/glog"
+	"github.com/mholt/archiver"
+	"github.com/pingcap/tidb-operator/cmd/backup-manager/app/constants"
+	"github.com/pingcap/tidb-operator/cmd/backup-manager/app/util"
+)
+
+// BackupOpts contains the input arguments to the backup command
+type BackupOpts struct {
+	Namespace   string
+	TcName      string
+	TidbSvc     string
+	Password    string
+	User        string
+	StorageType string
+	BackupName  string
+}
+
+func (bo *BackupOpts) String() string {
+	return fmt.Sprintf("%s/%s", bo.Namespace, bo.TcName)
+}
+
+func (bo *BackupOpts) getBackupFullPath() string {
+	return filepath.Join(constants.BackupRootPath, bo.getBackupRelativePath())
+}
+
+func (bo *BackupOpts) getBackupRelativePath() string {
+	backupName := fmt.Sprintf("backup-%s", time.Now().UTC().Format(time.RFC3339))
+	return fmt.Sprintf("%s_%s/%s", bo.Namespace, bo.TcName, backupName)
+}
+
+func (bo *BackupOpts) getDestBucketURI() string {
+	return fmt.Sprint("%s://%s%s", bo.StorageType, bo.getBackupRelativePath(), constants.DefaultArchiveExtention)
+}
+
+func (bo *BackupOpts) getTikvGCLifeTime(db *sql.DB) (string, error) {
+	var tikvGCTime string
+	row := db.QueryRow("select variable_value from ? where variable_name= ?", constants.TidbMetaTable, constants.TikvGCVariable)
```

Comment:
This has SQL injection problem.

Reply:
On the one hand, the input parameters of this SQL are controllable. In addition, parameterized statements use parameters instead of concatenating input variables in embedded SQL statements is safe. Refer to [here](https://github.com/astaxie/build-web-application-with-golang/blob/master/en/09.4.md#how-to-prevent-sql-injection)